### ISSUE TITLE
New "dependency" optimisation feature breaks "tfenv"

### ISSUE THREAD

**Issue by Fran-ELS-AMS** [gruntwork-io/terragrunt#1344]
The new `dependency` optimisation introduced in `terragrunt` version `0.23.35` breaks `tfenv` (at least on a Mac) when it's used with a `.terraform-version` file in the `terragrunt` root.

I think this is caused by the introduction of a temporary folder [here](https://github.com/gruntwork-io/terragrunt/pull/1311/files#diff-a8f51f0ba302e08284801441eab2c6e1R435-R443)

`terragrunt` changes its working directory to the temporary folder, and when it runs a `terraform` command from there `tfenv` cannot find the `.terraform-version` file that is in the project root, so it will use the system-wide version, which will often be either unset or a different version

This causes issues like this (`tfenv` system-wide version unset):

```
terragrunt plan
tg plan
[terragrunt] [</terragrunt/path>] 2020/09/14 11:35:11 Running command: terraform --version
[terragrunt] 2020/09/14 11:35:13 Terraform version: 0.12.29

...

[terragrunt] [</some/terragrunt/dependency/path] 2020/09/14 11:35:19 Running command: terraform output -json
Version could not be resolved (set by /usr/local/Cellar/tfenv/2.0.0/version or tfenv use <version>)
Version could not be resolved (set by /usr/local/Cellar/tfenv/2.0.0/version or tfenv use <version>)
Version could not be resolved (set by /usr/local/Cellar/tfenv/2.0.0/version or tfenv use <version>)
Version could not be resolved (set by /usr/local/Cellar/tfenv/2.0.0/version or tfenv use <version>)
Version could not be resolved (set by /usr/local/Cellar/tfenv/2.0.0/version or tfenv use <version>)
Version could not be resolved (set by /usr/local/Cellar/tfenv/2.0.0/version or tfenv use <version>)
Version could not be resolved (set by /usr/local/Cellar/tfenv/2.0.0/version or tfenv use <version>)
[terragrunt] 2020/09/14 11:35:22 exit status 1
```

or this (`tfenv` system-wide version different and incompatible):

```
terragrunt plan
tg plan
[terragrunt] [</terragrunt/path>] 2020/09/14 11:35:11 Running command: terraform --version
[terragrunt] 2020/09/14 11:35:13 Terraform version: 0.12.29

...

[terragrunt] [</some/terragrunt/dependency/path] 2020/09/14 11:35:19 Running command: terraform output -json
Failed to load state: state snapshot was created by Terraform v0.12.29, which is newer than current v0.12.26; upgrade to Terraform v0.12.29 or greater to work with this state
Failed to load state: state snapshot was created by Terraform v0.12.29, which is newer than current v0.12.26; upgrade to Terraform v0.12.29 or greater to work with this state
Failed to load state: state snapshot was created by Terraform v0.12.29, which is newer than current v0.12.26; upgrade to Terraform v0.12.29 or greater to work with this state
Failed to load state: state snapshot was created by Terraform v0.12.29, which is newer than current v0.12.26; upgrade to Terraform v0.12.29 or greater to work with this state
Failed to load state: state snapshot was created by Terraform v0.12.29, which is newer than current v0.12.26; upgrade to Terraform v0.12.29 or greater to work with this state
Failed to load state: state snapshot was created by Terraform v0.12.29, which is newer than current v0.12.26; upgrade to Terraform v0.12.29 or greater to work with this state
Failed to load state: state snapshot was created by Terraform v0.12.29, which is newer than current v0.12.26; upgrade to Terraform v0.12.29 or greater to work with this state
[terragrunt] 2020/09/14 11:38:02 exit status 1
```

Please note how `terragrunt` outputs the `terraform` version at the beginning (the correct version), when the working directory is still the same directory in which I'm running `terragrunt plan`, but everything changes when it's working on the dependencies, and `tfenv` suddenly doesn't know the `terraform` version, or it knows a different one

**Comment by Fran-ELS-AMS** [comment]
P.S.:
This can be worked-around by setting the system-wide `terraform` version for `tfenv` with:

```tfenv use <terraform_version>```

or by setting the new `terragrunt` variable `disable_dependency_optimization` (introduced in `terragrunt` version `0.23.37`) to `true` in the `remote_state` block.

Both the above work-arounds are not acceptable as solutions.
The first one is not the best way to use `tfenv` and it makes having a `.terraform-version` file useless.
The second one disables the `dependency` optimisation, which was indeed a great and much appreciated optimisation.

**Comment by Fran-ELS-AMS** [comment]
Maybe changing [this line](https://github.com/gruntwork-io/terragrunt/pull/1311/files#diff-a8f51f0ba302e08284801441eab2c6e1R435) to one of the below could be a solution, @yorinasub17 (tagging you since you seem to be the author of the change)?

```
tempWorkDir, err := ioutil.TempDir("./", ".*")
```

or

```
tempWorkDir, err := ioutil.TempDir("./.terragrunt-cache", "")
```

or anything similar to create the temporary working directory still inside the `terragrunt` project

**Comment by yorinasub17** [comment]
Yup that makes sense, although using `./.terragrunt-cache` directly is not correct if the terragrunt cache is configured to a different directory. Will try to work on this this week.

**Comment by yorinasub17** [comment]
Just opened a PR with a potential fix. Can you see if that works for you? I released this as https://github.com/gruntwork-io/terragrunt/releases/tag/v0.24.2-alpha.1 to make it easy to get the binary.

**Comment by Fran-ELS-AMS** [comment]
> Just opened a PR with a potential fix. Can you see if that works for you? I released this as https://github.com/gruntwork-io/terragrunt/releases/tag/v0.24.2-alpha.1 to make it easy to get the binary.

It seems to be working perfectly.

I've tested with:
- the system-wide `terraform` version for `tfenv` **unset**
- the system-wide `terraform` version for `tfenv` **set to a version that's incompatible with the current version of `terragrunt` (`0.12.26`)**

Before each test I made sure to remove all `.terragrunt-cache` folders with:
```
find . -type d -name ".terragrunt-cache" -prune -exec rm -rf {} \;
```
Both tests were successful.

I then re-run both tests without removing the `.terragrunt-cache` folders, just in case, to test both cases.

Both tests were successful again.

The temporary directories, containing only the `backend.tf` file, get created (one each) in the `.terragrunt-cache` folder of each dependency for which `terragrunt` runs `terraform output -json` and it almost immediately gets removed, leaving behind an empty `.terragrunt-cache` folder.

LGTM
